src/broker/controller.c

```diff
@@ -83,6 +83,56 @@ int controller_name_activate(ControllerName *name) {
         return controller_dbus_send_activation(name->controller, name->path);
 }
 
+static int controller_reload_compare(CRBTree *t, void *k, CRBNode *rb) {
+        ControllerReload *reload = c_container_of(rb, ControllerReload, controller_node);
+        uint32_t serial = *(uint32_t *)k;
+
+        if (serial < reload->serial)
+                return -1;
+        if (serial > reload->serial)
+                return 1;
+
+        return 0;
+}
+
+ControllerReload *controller_reload_free(ControllerReload *reload) {
+        if (!reload)
+                return NULL;
+
+        c_rbtree_remove_init(&reload->controller->reload_tree, &reload->controller_node);
+        free(reload);
+
+        return NULL;
+}
+
+static int controller_reload_new(ControllerReload **reloadp, Controller *controller) {
+        CRBNode **slot, *parent;
+        ControllerReload *reload;
+        uint32_t serial = ++controller->serial;
+
+        slot = c_rbtree_find_slot(&controller->reload_tree, controller_reload_compare, &serial, &parent);
+        assert(slot);
```

Comment:
Ugh. I hate doing this on 32bit variables. Can't we make this a loop? Increment serial, look it up, continue if taken? Yeah yeah, trusted peer and so on... but still? Dunno..

Reply:
I agree, let's not rely on this being privileged.